akka-stream/src/main/scala/akka/stream/impl/ActorRefBackpressureSinkActor.scala

```diff
@@ -0,0 +1,108 @@
+
+/**
+ * INTERNAL API
+ */
+private[akka] class ActorRefBackpressureSinkStage[Out](ref: ActorRef, onInitMessage: Any,
+                                                       ackMessage: Any,
+                                                       onCompleteMessage: Any)
+  extends SinkStage[Out, Unit]("ActorRefBackpressureSink") {
+
+  val maxBuffer = module.attributes.getAttribute(classOf[InputBuffer], InputBuffer(16, 16)).max
+  require(maxBuffer > 0, "Buffer size must be greater than 0")
+
+  override def createLogicAndMaterializedValue: (GraphStageLogic, Unit) = {
+
+    val buffer: util.Deque[Out] = new util.ArrayDeque[Out]()
+    var acknowledgementReceived = false
+    var completeReceived = false
+    var pullRequestIsSent = true
+
+    val logic = new GraphStageLogic(shape) {
+
+      private val callback: AsyncCallbackNoArgs = getAsyncCallback(() ⇒ {
+        if (!buffer.isEmpty) sendData()
+        else acknowledgementReceived = true
+      })
+
+      private val deathWatchCallback: AsyncCallbackNoArgs =
+        getAsyncCallback(() ⇒ completeStage())
+
+      implicit val internalRefCallback = new akka.actor.MinimalActorRef {
+        override def provider: ActorRefProvider =
+          throw new UnsupportedOperationException("ActorRefBackpressureSink MinimalActor does not provide")
+        override def path: ActorPath =
+          ActorMaterializer.downcast(interpreter.materializer).supervisor.path / "ActorRefWithAckMinimalActor"
+        override def !(message: Any)(implicit sender: ActorRef = Actor.noSender): Unit = {
+          message match {
+            case msg if msg == ackMessage ⇒ callback.invoke()
+            case _                        ⇒ //ignore all other messages
+          }
+        }
+        override def sendSystemMessage(message: SystemMessage): Unit = message match {
+          case DeathWatchNotification(_, _, _) ⇒ deathWatchCallback.invoke()
+          case _                               ⇒ //ignore all other messages
+        }
+      }
+
+      override def preStart() = {
+        val internalRef = ref.asInstanceOf[InternalActorRef]
+        internalRef.sendSystemMessage(Watch(internalRef, internalRefCallback))
+
+        ref ! onInitMessage
+        pull(in)
+      }
+
+      private def sendData(): Unit = {
+        if (!buffer.isEmpty) {
+          ref ! buffer.poll()
+          acknowledgementReceived = false
+        }
+        if (buffer.isEmpty && completeReceived) finish()
+      }
+
+      private def finish(): Unit = {
+        ref ! onCompleteMessage
+        completeStage()
+      }
+
+      private def sendPullIfAllowed(): Unit =
+        if (!pullRequestIsSent) {
+          pullRequestIsSent = true
+          pull(in)
+        }
+
+      setHandler(in, new InHandler {
+        override def onPush(): Unit = {
+          buffer offer grab(in)
+          pullRequestIsSent = false
+          if (acknowledgementReceived) sendData()
+          if (buffer.size() < maxBuffer) sendPullIfAllowed()
```

Comment:
it's allowed, I agree. But we cannot request more than buffer size (as waiting for actorRef acknowledgment)